Guard changes

chiragsibm opened the pull request:
This commit contains the changes related in libguard

Commit:
libguard: Adding space check before creating record

Changes:
-Adding check in create API, which will calculate the available
size in the file before writing the guard record.

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

Commit:
libguard: Add support for guard record history

Changes:
-Currently in libguard the guard records were wiped out
from the file, but changing it now to maintain the record
history. Hence instead of deleting, making the recordID
as FFFFFFFF and not displaying it.
-Change in create guard function, for marking correct recordId
because of change in delete function as we are marking recordId
as FF..FF it will take maxId is that and will not allocate correct
recordId to newly created guard record.
-Allowing to create guard of resolved guard record
-Adding function invalidateAll to invalidate all the guard records
i.e. mark recordId as FFFFFFFF for all the guard records present.
-Changing clearAll function to erase the data in guard file.
-Adding UT for invalidateAll function.
-Changes in Guard tool to not display invalidated guard records.

Test:
-Created guard record deleted the same, the recordId was setting
to FFFFFFFF and not getting displaying in guard tool output.
-Clearall is marking all recordId as FF...FF
-Created a guard record and deleted the same, tried to guard
the same record.

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

Commit:
libguard: Storing oldest resolved record's position

Change:
-Adding check in create api to store the location/position
of invalid/resolved record, incase there is no space left
that resolved record position will be used to create a
new record.
-Adding UT for the above change where when no space is
left in guard file the oldest guard records will be replaced
with the new guard record.

Test:
-Created guard records and deleted few of them,
for the deleted guard records record id got set
to 0xFFFFFFFF. When no space is left in guard file
the 1st resolved record location will be used to create
the new record.

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -169,6 +188,9 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
     }
 
+	// Space left in file before writing new record
+    fileSizeCheck(file.size(), lastPos + 1);
```
can you delete resolved record (invalid id) when you don't have space?
OR the customer need to trigger deleteAll? If doing the deleteAll then it will delete all or only resolved record?

Need some documentation about what action customer needs to do or libguard won't do such an activity and the consumer of the libguard needs to decide when the guard file is overflowing?

chiragsibm replied:
We can delete the invalid guard records. If we are handling this at libguard end than no need of throwing exception saying there is no space left in file. We need to decide on this.

RameshIyyar replied:
- Where you have documented that information? what is the action you won't do when a file is fulled so that consumers can decide what needs to be done? - I think we cannot say what need by done by the caller, its up to them/

> If we are handling this at libguard end than no need of throwing

Why so, it will drive based on what you are going to the delete operation let say, the file is fulled only with unresolved records - what you will do ?

> We need to decide on this.

Agree, Bring into some forum.

chiragsibm replied:
As per discussion, will delete the invalid/resolved guard records when space is full.

dhruvibm replied:
I think you should be deleting only one resolved guard record to add a new one

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,25 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
+ * @param[in] count - No of records present
+ *
+ */
+void fileSizeCheck(uint32_t size, uint32_t count)
```
size parameter type should be size_t right?
If so, Please make sure all the places

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,25 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
+ *
+ * @param[in] size - File size
+ * @param[in] count - No of records present
+ *
```
Need documentation about return (you are throwing exception)

chiragsibm replied:
Will do this.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,25 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
+/**
+ * @brief Function to get the size remaining in the file
+ *
+ * @param[in] size - File size
```
Guard file size ?

chiragsibm replied:
Yes guard file size.

RameshIyyar replied:
I am asking can we you change "File size" into "Guard File Size" - I feel file size is common name and it will give confusion right?

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,25 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
 }
 
+/**
+ * @brief Function to get the size remaining in the file
```
guard file?

chiragsibm replied:
Yes guard file. You want me to change it to guard file ?

RameshIyyar replied:
yes to avoid confusion

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -169,6 +188,9 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
         lastPos++;
     }
 
+	// Space left in file before writing new record
```
a new record?

chiragsibm replied:
Yes, before creating new guard record we are checking for the space in guard file.

RameshIyyar replied:
> Space left in file before writing new record

Space left in file before writing new a record

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,25 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
+ * @brief Function to get the size remaining in the file
+ *
+ * @param[in] size - File size
+ * @param[in] count - No of records present
```
No of records present in the guard file?

chiragsibm replied:
Yes, will change it.

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,25 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
+    if (aval_size < sizeof(record))
+    {
+        guard_log(GUARD_ERROR, "Size left in file is %d\n", aval_size);
+        throw std::runtime_error("Enough size is not available in guard file");
```
What exception you will throw in this case? The list which you shared is covered this case?

chiragsibm replied:
Yes its covered.  guardMaxSizeReached

RameshIyyar commented on guard.cpp:
```diff
@@ -18,7 +18,8 @@ void guardList()
     for (const auto& elem : records)
     {
-        if (elem.errType != GARD_Reconfig)
+        if ((elem.errType != GARD_Reconfig) &&
```
guard tool does not have any requirement to show the invalid record (resolved record)

chiragsibm replied:
We should not show any invalid recordId in tool. Even in reconfig path the guard recordId is set to FFFFFFFF. And displaying an invalid record ID doesnt make much sense.

RameshIyyar replied:
Reconfig and Resolved (Invalid) are different right?
It will be greater if you gave the option to show resolved records but, it's your wish.

chiragsibm replied:
The issue is for reconfig type guard record the recordId gets updated to FF..FF. Not sure if we should think as its resolved ?

RameshIyyar replied:
I am not talking for reconfig record (fake guard record), just checking whether the tool has a requirement to show the resolved record (0xFFFFFFFF)?
If no, do you need to check with the tool user?

chiragsibm replied:
We will show the resolved guard records. Added option for that.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -153,7 +153,6 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
     //! check if guard record already exists
     int pos = 0;
     int lastPos = 0;
-    uint32_t maxId = 0;
```
Why you are making changes in create() API? no information about this in the commit message?

chiragsibm replied:
Will add in the commit msg. basically this required as we are not deleting the guard record we are just marking the recordID as FFFFFFF so the maxid will get set to that and we will not get the actual count present.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -197,7 +191,7 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
         offset = lastPos * sizeof(guard);
         memset(&guard, 0xff, sizeof(guard));
-        guard.recordId = htobe32(maxId + 1);
+        guard.recordId = htobe32(lastPos + 1);
```
why need to do this?

chiragsibm replied:
Explained above.

RameshIyyar commented on libguard/include/guard_record.hpp:
```diff
@@ -12,6 +12,7 @@ namespace guard
 #define GUARD_MAGIC "GUARDREC" //! "GUARDREC"
 /* From hostboot: src/include/usr/hwas/hwasPlatDeconfigGard.H */
 const uint8_t CURRENT_GARD_VERSION_LAYOUT = 0x1;
+#define INVALID_RECORD_ID 0xFFFFFFFF
```
GUARD_RESOLVED?

chiragsibm replied:
Can discuss on name.

RameshIyyar replied:
Bring it in some forum if you have concern

chiragsibm replied:
Changing to GUARD_RESOLVED.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -329,6 +329,36 @@ void clearAll()
     }
 }
 
+void clearInvalidRecords()
```
is it invalid or resolved?

chiragsibm replied:
I am not sure why we can call it as resolved or not. Basically I have seen that in reconfig look hostboot sets the record ID as F..F, so not sure to call it resolved.

RameshIyyar replied:
I don't know the intention of updating ID as `0xFFFFFFFF` - I heard like if guarded hardware is repair or replaced then id will get change as `0xFFFFFFFF` so, we are saying it as resolved. if you have seen some other use cases then bring in some forum or pick whatever makes sense to you.

chiragsibm replied:
Will change to resolved.

RameshIyyar commented on libguard/guard_interface.hpp:
```diff
@@ -63,6 +63,14 @@ void clear(const uint32_t recordId);
+ *
+ * @return NULL
+ */
+void clearInvalidRecords();
```
Do you need this support in the guard tool?

chiragsibm replied:
Need to discuss.

RameshIyyar reviewed (commented):
- Reviewed last three commits

Please make sure which name is best resolved vs invalid

Commit:
libguard: Adding user defined exceptions

Changes:
-Adding user defined exceptions in libguard so that the caller
can distinguish the type of error at his end.
-Changes in respective UTs.

Test:
Created guard record, tried creating the same.
Deleting a non guarded record.
Ran UTs in docker env.

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+    {
+    }
+
+    virtual const char* what() const noexcept override
```
Do you need virtual here?

chiragsibm replied:
Note required removed. Done

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+#include <exception>
+#include <string>
+
+class guardFileOpen : public std::exception
```
UpperCamelCase?
GuardFileOpenFailed?

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+class guardFileOpen : public std::exception
+{
+  public:
+    explicit guardFileOpen(const std::string& message) : message(message)
```
Why you are expecting a string from a thrower? Why can not use that failure reason in this class itself? or is it will vary from one place to another?

chiragsibm replied:
For few I can add in class itself, but for some we have different strings

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+    {
+        return message.c_str();
+    }
+
```
Do you think adding the below function will help more to catcher?
- name() - it can return the name of the exception? "GuardFileOpen"?
- description() - it can return the description of the exception? 
- what() - it can return both i.e name and description?

chiragsibm replied:
I am unable to get what name i.e. function name will do at user end.

RameshIyyar replied:
They will know the name of the exception and do appropriate action but, if you introduce a base exception for the guard i.e for the selected guard exception catcher can do whatever he needs.

chiragsibm replied:
what extra actions are they planning to do as a part of exception handling? it makes sense only if there is a common action that needs to be done for all types of exception. and then there are extra specific actions that needs to be done for each exception.

RameshIyyar replied:
>if there is a common action that needs to be done for all types of exception. and then there are extra specific actions that needs to be done for each exception.

Yes but, as a library, we can not say that restriction - The caller can do anything if he wants so our library supports it. 
If you are planning to do when someone requesting then I am fine - Just giving concern as Reviewer from library context.

chiragsibm replied:
Will do when required.

ojayanth replied:
Any value add in introducing custom exception here , This will be useful for user incase need to share the error code or additional info .  Instead of making multiple exception  you can create 
struct error_info                                                               
{                                                                               
   ERR_TYPE err_type;                                                           
   const char *msg;                                                             
  void const*  private_data;                         
}; 

This will help to catch user generic exception and get require info , also incase neet add any additional data in future it is easy.

ojayanth replied:
Please create todo in guard repository and add TOD number here , with proposed details.

chiragsibm replied:
Created issue https://github.com/ibm-openbmc/dev/issues/3325

ojayanth replied:
please create issue in guard repository

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+#include <exception>
+#include <string>
+
+class guardFileOpen : public std::exception
```
Why not have a common base class for all your guard exceptions? 
So, that catcher no need to worry about write multiple catches if he going to take some common action?

still, you can argue why not caller use std::exception? this is a standard exception and the catcher maybe do a special action for all guard exceptions and different activities for standard exception

chiragsibm replied:
In that case how will user differentiate between the failures ?

ojayanth replied:
> Why not have a common base class for all your guard exceptions?
> So, that catcher no need to worry about write multiple catches if he going to take some common action?
> 
> still, you can argue why not caller use std::exception? this is a standard exception and the catcher maybe do a special action for all guard exceptions and different activities for standard exception

+1

ojayanth replied:
> In that case how will user differentiate between the failures ?

you can define custon structre , which includes the required error data.

struct error_info                                                               
{                                                                               
   ERR_TYPE err_type;                                                           
   const char *msg;                                                             
   void const*  private_data;                         
};

chiragsibm replied:
Not really this will lead to more changes as compared to adding exceptions, as we need to change the return type of every function also changes in tool side. Also we thought of opting for CPP way and writing custom exceptions.

chiragsibm replied:
Did the inheritance(i.e. common base class) part but got some issues with multiple arguments so for now going with inheritance changes only, struct approach need more changes and UT changes as well. If required in future will change the infra to accommodate more args.

RameshIyyar reviewed (commented):
- Please add UT for all exceptions (wherever is possible).
- Please document the exception usage like what is the scope of that exception and conditions that need to meet to use that exception to throw.
- I have added comments for one exception it applicable to all exceptions.
- I would spit patches differently.
  - Add a separate patch for all exceptions, good documentation with the condition, details in the commit message, and UT (if possible).
  - Add a patch to throw each exception (more than one throw is possible for the same exception) from libguard and UT (if possible) - it help to reviewer to ask - is there anything to consider for that exception from existing code.

RameshIyyar commented:
Please rebase your patch before pushing the next patch set for this pr

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+    std::string message;
+};
+
+class guardFileRead : public std::exception
```
GuardFileReadFailed?

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+    std::string message;
+};
+
+class guardFileWrite : public std::exception
```
GuardFileWriteFailed?

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+    std::string message;
+};
+
+class guardInvalidEntity : public std::exception
```
InvalidEntityPath?

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+    std::string message;
+};
+
+class guardRecordExist : public std::exception
```
AlreadyGuarded?

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+    std::string message;
+};
+
+class guardMaxSizeReached : public std::exception
```
GuardFileOverFlowed?

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,100 @@
+
+#include <exception>
+#include <string>
+
```
namespace ?
```
namespace openpower
{
namespace guard
{
namespace exception
{
```

Commit:
tool: Changes in options of guard tool

Changes:
-Making options of guardtool more meaningful as per the
backend functionality.
Like -d to -i, this will Invalidate a single Guard record
Changing --clear-all to --reset, which will erase all the
guard records.
-Adding -I, will Invalidates all the Guard records
-Adding -a option to list the resolved guard records
-Updating README with new options
-Setting GUARD_VERSION in meason.build to return correct value.

Test:
-creating and invalidating guard records, checking if its getting
displayed with new option or not.
-with changed options able to invalidate the guard records.

[==========] Running 15 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 15 tests from TestGuardRecord
[ RUN      ] TestGuardRecord.CreateGuardRecord
[       OK ] TestGuardRecord.CreateGuardRecord (0 ms)
[ RUN      ] TestGuardRecord.ClearGuardGoodPathTest
[       OK ] TestGuardRecord.ClearGuardGoodPathTest (0 ms)
[ RUN      ] TestGuardRecord.DeleteGuardGoodPathTest
[       OK ] TestGuardRecord.DeleteGuardGoodPathTest (0 ms)
[ RUN      ] TestGuardRecord.NegTestCaseEP
[       OK ] TestGuardRecord.NegTestCaseEP (0 ms)
[ RUN      ] TestGuardRecord.NegTestCaseFullGuardFile
Guard file size is 656 and space remaining in GUARD file is 0

[       OK ] TestGuardRecord.NegTestCaseFullGuardFile (0 ms)
[ RUN      ] TestGuardRecord.AlreadyGuardedTC
Already guard record is available in the GUARD partition
[       OK ] TestGuardRecord.AlreadyGuardedTC (0 ms)
[ RUN      ] TestGuardRecord.GetCreatedGuardRecordTC
[       OK ] TestGuardRecord.GetCreatedGuardRecordTC (0 ms)
[ RUN      ] TestGuardRecord.DeleteByEntityPath
[       OK ] TestGuardRecord.DeleteByEntityPath (0 ms)
[ RUN      ] TestGuardRecord.DeleteWithNotExistentEntity
Guard record not found
[       OK ] TestGuardRecord.DeleteWithNotExistentEntity (0 ms)
[ RUN      ] TestGuardRecord.DeleteByRecordId
[       OK ] TestGuardRecord.DeleteByRecordId (0 ms)
[ RUN      ] TestGuardRecord.DeleteWithNotExistentRecordId
Guard record not found
[       OK ] TestGuardRecord.DeleteWithNotExistentRecordId (0 ms)
[ RUN      ] TestGuardRecord.GetGuardFilePathTC
[       OK ] TestGuardRecord.GetGuardFilePathTC (0 ms)
[ RUN      ] TestGuardRecord.GetGuardFilePathWhenLibguradDidNotInitTC
Guard file is not initialised.
[       OK ] TestGuardRecord.GetGuardFilePathWhenLibguradDidNotInitTC (0 ms)
[ RUN      ] TestGuardRecord.ClearGuardInvalidatePathTest
[       OK ] TestGuardRecord.ClearGuardInvalidatePathTest (0 ms)
[ RUN      ] TestGuardRecord.ClearResolvedGuardRecord
[       OK ] TestGuardRecord.ClearResolvedGuardRecord (0 ms)
[----------] 15 tests from TestGuardRecord (4 ms total)

[----------] Global test environment tear-down
[==========] 15 tests from 1 test suite ran. (5 ms total)
[  PASSED  ] 15 tests.
------------------------------------------------------------------------------
Ok:                 1
Expected Fail:      0
Fail:               0
Unexpected Pass:    0
Skipped:            0
Timeout:            0

Signed-off-by: Chirag Sharma <chirshar@in.ibm.com>

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,28 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
+ *         Throw exception, if GUARD file doesn't have space to add
+ *         new guard record.
+ */
+void fileSizeCheck(size_t size, uint32_t count)
```
Do we need this function as an interface?  From your below commit message, looks like it was used for internal validation before create. - is not like that?
```
-Adding fileSizeCheck, which will calculate the available
size in the file before writing the guard record.
```
OR Your intention is before to create guard record user check whether space is there or not? If so, How users can get to know, how many records can add by using returned size? since he doesn't know the record size or he need to do that math at his side?

I am wondering what users can do with this interface since you are going to throw an exception for that in a later patch right?

IIFC, I think I have requested you return the error code if space is full and add a trace of how many records can add in a trace OR give the interface to get to know the no of records allowed to add so that a user can exchange that information to an external user or add trace at his side (IMO, In OpenBMC DBus we don't care about no of entries) - Correct me If I am wrong.

RameshIyyar replied:
Please clarify the above comments - For me, looks it is used for internal purpose.

chiragsibm replied:
It is just for internal purpose. Very function we introduce in interface doesnt get exported to external user. This function is an internal check before creating guard record.

RameshIyyar replied:
>It is just for internal purpose.

Then Please move into .cpp and mark it as a `static` function 

> Very function we introduce in interface doesnt get exported to external user.

I would say modify those as well if you want. 

In general, the library functions are not intended for the user than those functions need to be static in the CPP unit file and if more than one CPP unit needs then make as utils and we should no install that utils header file so that the user won't aware or he knows it is a utility function for the library.

chiragsibm replied:
Its already added in guard_interface.cpp. Changed to static function.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -323,6 +323,36 @@ void clearAll()
+
+    for_each_guard(file, pos, guard)
+    {
+        if (guard.recordId != 0xffffffff)
```
why not using macro which you defined to use `0xffffffff`

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -315,6 +317,7 @@ void clearAll()
     {
         for_each_guard(file, pos, existGuard)
         {
+		    printf("pos %d, lastPos %d\n", pos, lastPos);
```
is it required? you said added for debug purpose and going to remove but still seeing the trace.

RameshIyyar commented on guard.cpp:
```diff
@@ -18,31 +18,63 @@ void guardList()
     {
-        if ((elem.errType != GARD_Reconfig) &&
-            (elem.recordId != GUARD_RESOLVED))
+        if (displayResolved)
```
Please simplify the code both if and else doing the same only condition is different.

RameshIyyar replied:
Please clarify this

chiragsibm replied:
Already simplified the code till extend possible. Both if and else condition are different.

RameshIyyar replied:
What I think we can do like below
```
void guardList(bool displayResolved)
{
    auto records = getAll();
    if (!records.size())
    {
        std::cout << "No Records to display" << std::endl;
        return;
    }

    bool isHeaderPrinted = false;
    for (const auto& elem : records)
    {
        // Don't print Reconfig (aka Fake record which is used for internal
        // purpose by BMC and Host application)
        if (elem.errType == GARD_Reconfig)
        {
            continue;
        }
        // User interested to see only resolved records
        else if (displayResolved && (elem.recordId != GUARD_RESOLVED))
        {
            continue;
        }
        // User interested to see only unresolved records
        else if (!displayResolved && (elem.recordId == GUARD_RESOLVED))
        {
            continue;
        }

        // Don't print the header if already printed since records mixed
        // of resolved and unresolved records so if have only either one
        // in retrieved records and user tried to see opposite one then
        // we should not print the header else the user will get confused. 
        if (!isHeaderPrinted)
        {
            std::cout << "ID       | ERROR    |  Type  | Path " << std::endl;
            isHeaderPrinted = true;
        }

        std::cout << std::hex << std::setw(8) << std::setfill('0')
                  << elem.recordId;

        std::cout << " | ";
        std::cout << std::hex << std::setw(8) << std::setfill('0')
                  << elem.elogId;

        std::cout << " | ";
        std::optional<std::string> gReasonToStr =
                  guardReasonToStr(elem.errType);
        std::cout << *gReasonToStr;

        std::cout << " | ";
        std::optional<std::string> physicalPath =
                  getPhysicalPath(elem.targetId);
        if (!physicalPath)
        {
            std::cout << "Unknown ";
        }
        else
        {
            std::cout << *physicalPath;
        }
        std::cout << std::endl;
    }

    if (!isHeaderPrinted)
    {
        std::cout << "No " << (displayResolved ? "resolved" : "unresolved")
                  << " records to display" << std::endl;
    }
}
```

Did you see any issue with this?

RameshIyyar commented on guard.cpp:
```diff
@@ -104,6 +143,10 @@ int main(int argc, char** argv)
         app.add_flag("-r, --clearall", clearAll,
                      "Clears GUARD states for all resources");
+        app.add_flag("-s, --clearresolvedrecords", clearResolvedRecords,
+                     "Clears GUARD states for resolved resources");
```
are deleting or clearing state?

chiragsibm replied:
Deleting. Done

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -139,19 +143,31 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
     {
-        guard_log(GUARD_ERROR, "Size left in GUARD file is %d\n", aval_size);
-        throw GuardFileOverFlowed("Enough size is not available in GUARD file");
+        clearResolvedRecords();
```
It will be great If you added some comments on why you are doing this - It will help for future reference.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -183,7 +183,8 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
         {
-            if (existGuard.errType == GARD_Reconfig)
+            if ((existGuard.errType == GARD_Reconfig) ||
+                (existGuard.recordId == GUARD_RESOLVED))
```
Must be a new record right? Why overriding the existing record? How you will get the resolved record history if you override the existing record?

chiragsibm replied:
Done. As discussed and agreed creating new guard record

RameshIyyar commented on guard.cpp:
```diff
@@ -9,13 +9,32 @@ using namespace openpower::guard;
 
 void guardList(bool displayResolved)
 {
+    bool display = false;
```
> -When there is no resolved guard record to display, the header
gets displayed. So in this case displaying message
"No Records to display"

When there is only resolved guard records to display, the header
gets displayed. So in this case displaying a message
"No Records to display"

RameshIyyar commented on guard.cpp:
```diff
@@ -9,13 +9,32 @@ using namespace openpower::guard;
         return;
     }
-    std::cout << "ID       | ERROR    |  Type  | Path " << std::endl;
+    for (const auto& elem : records)
```
why you need two for loop? you can simplify the code by using stringstream

chiragsibm replied:
Simplified. Done

RameshIyyar commented on README.md:
```diff
@@ -41,6 +41,8 @@ Options:
   -d,--delete TEXT       Delete GUARD record, expects physical path as input.
   -l,--list              Listing of GUARDed resources.
   -r,--clearall          Clears GUARD states for all resources.
+  -s,--clearresolvedrecords   Clears GUARD states for resolved resources
```
This should go in the respective commit, not a new commit.

RameshIyyar commented on test/guard_intf_test.cpp:
```diff
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: Apache-2.0
 #include "libguard/guard_common.hpp"
 #include "libguard/guard_entity.hpp"
+#include "libguard/guard_exception.hpp"
```
Please add in the respective commit

RameshIyyar commented on test/guard_intf_test.cpp:
```diff
@@ -271,3 +275,27 @@ TEST_F(TestGuardRecord, GetGuardFilePathWhenLibguradDidNotInitTC)
     EXPECT_EQ(retGuardFilePath, guardFile);
 }
+
+TEST_F(TestGuardRecord, ClearResolvedGuardRecords)
```
Please add in the respective commit

dhruvibm commented on libguard/guard_interface.cpp:
```diff
@@ -124,6 +124,28 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
+void fileSizeCheck(size_t size, uint32_t count)
+{
+    GuardRecord record;
+    uint32_t aval_size = size - (count * sizeof(record) + headerSize);
```
file size is fixed right?

chiragsibm replied:
Yes, its fixed but in case it changes in future that's why getting the file size on fly.

dhruvibm commented on guard.cpp:
```diff
@@ -18,7 +18,8 @@ void guardList()
     {
-        if (elem.errType != GARD_Reconfig)
+        if ((elem.errType != GARD_Reconfig) &&
+            (elem.recordId != GUARD_RESOLVED))
```
we may need to add another option to show even resolved guard record to show the history

chiragsibm replied:
Its already added i.e. -p option will display the resolved guard records. Just that i am handling both guarded and resolved resources in one function.

ojayanth replied:
This option is not included in in the guard help?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -247,55 +248,22 @@ GuardRecords getAll()
-    }
-}
-
 void clear(const EntityPath& entityPath)
```
JS: Looks clear(const EntityPath& entityPath) and clear(const uint32_t recordId) are same except the record lookup inputs i.e entitypath or id - you can simplify the code if you like to do that.

chiragsibm replied:
Will take this later.

ojayanth replied:
This is valid command , just a copy of function, and log message also same here.

chiragsibm replied:
Will take this in separate PR.

ojayanth replied:
since this is new function, should handle in this PR only.

chiragsibm replied:
This function is from last PR. I didn't introduced this function in this PR. 
I told I will take care of making a common code later instead of handling in this PR.

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -247,55 +248,22 @@ GuardRecords getAll()
+            offset = lastPos * sizeof(existGuard);
+            existGuard.recordId = GUARD_RESOLVED;
+            file.write(offset + headerSize, &existGuard, sizeof(existGuard));
             found = true;
```
After deleting the record also, why you are looping?

chiragsibm replied:
Done

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -306,23 +274,24 @@ void clear(const EntityPath& entityPath)
+            offset = lastPos * sizeof(existGuard);
+            existGuard.recordId = GUARD_RESOLVED;
+            file.write(offset + headerSize, &existGuard, sizeof(existGuard));
             found = true;
```
After deleting the record also, why you are looping?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -247,55 +248,22 @@ GuardRecords getAll()
         if (existGuard.targetId == entityPath)
         {
-            delpos = pos;
+            offset = lastPos * sizeof(existGuard);
```
why lastPos and offset is required? since `for_each_guard` returning the position of readed record right?

chiragsibm replied:
offset we require to go to that particular record location. lastPos is not required we can use pos instead.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -306,23 +274,24 @@ void clear(const EntityPath& entityPath)
         if (be32toh(existGuard.recordId) == recordId)
         {
-            delpos = pos;
+            offset = lastPos * sizeof(existGuard);
```
why lastPos and offset is required? since `for_each_guard` returning the position of readed record right?

chiragsibm replied:
Explained above. will use pos instead of lastPos

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -333,41 +302,30 @@ void clear(const uint32_t recordId)
-            elem.recordId = htobe32(pos + 1);
-            file.write(offset + headerSize, &elem, sizeof(guard));
-            pos++;
+            offset = lastPos * sizeof(existGuard);
```
why lastPos and offset is required? since `for_each_guard` returning the position of readed record right?

chiragsibm replied:
offset is for readability pos will return like the position i.e. 0,1,2 etc. lastPos is not required can use pos.

RameshIyyar commented on libguard/guard_interface.hpp:
```diff
@@ -63,6 +63,14 @@ void clear(const uint32_t recordId);
 void clearAll();
 
+/**
+ * @brief Clearing old guard record with recordId
```
- Clearing a first record that is resolved from the guard records list?
- Please change the function name as `clearFirstResolvedRecord`
- is it required to be an external interface? - is it for internal purpose right?

chiragsibm replied:
We are exporting to guard tool. Will change the name. Yes it will clear the 1st/oldest resolved guard record from the list.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -330,6 +330,43 @@ void clearAll()
 }
 
+void clearResolvedRecords()
+{
```
Too many variables :) - If possible, please optimize or add a comment on what you are doing - it's a pain to understand the complex logic.

chiragsibm replied:
done

RameshIyyar commented on test/guard_intf_test.cpp:
```diff
@@ -271,3 +271,27 @@ TEST_F(TestGuardRecord, GetGuardFilePathWhenLibguradDidNotInitTC)
+TEST_F(TestGuardRecord, ClearResolvedGuardRecords)
+{
+    openpower::guard::libguard_init();
+    std::string phyPath = "/sys-0/node-0/dimm-0";
```
why you need variable for the physical path you can pass directly to the function right?

RameshIyyar replied:
Are you planning to address this? If not, please clarify

chiragsibm replied:
Yes, we can directly pass the string i.e. phy path to function but for readability purpose or say for more simpler format, I am adding it in a variable  and passing the same.

RameshIyyar commented on test/guard_intf_test.cpp:
```diff
@@ -271,3 +271,27 @@ TEST_F(TestGuardRecord, GetGuardFilePathWhenLibguradDidNotInitTC)
+    phyPath = "/sys-0/node-0/dimm-2";
+    entityPath = openpower::guard::getEntityPath(phyPath);
+    openpower::guard::create(*entityPath);
+    phyPath = "/sys-0/node-0/dimm-2";
```
What you are achieving here you can use the existing entity path which got for same physical path right?
same for next physical path as well

chiragsibm replied:
Done

RameshIyyar commented on libguard/guard_file.cpp:
```diff
@@ -22,22 +24,28 @@ GuardFile::GuardFile(const fs::path& file) : guardFile(file)
         fileSize = file.tellg();
     }
-    catch (const std::exception& ex)
+    catch (const GuardFileOpenFailed& ex)
```
- Already you are throwing exceptions in the above so why catching and throwing the same exception? 
- Also too many traces are possible for same exception

chiragsibm replied:
Done.

RameshIyyar commented on README.md:
```diff
@@ -34,14 +34,16 @@ GUARD Tool
+  -d,--delete TEXT         Delete GUARD record, expects physical path as input.
+  -l,--list                Listing of GUARDed resources.
+  -r,--clearall            Clears GUARD states for all resources.
+  -s,-clearresolvedrecord  Delete GUARD states for resolved resource
```
You are deleting only one first record which is resolved so I am not using what the user will do with this

OR

Add description like what the user can expect by this comment.

chiragsibm replied:
Will change to "Delete GUARD states for oldest resolved resource". 
Mainly for FVT to test the function as they can't access the libguard directly.. So this option is added.

chiragsibm replied:
Added a note.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -194,6 +194,11 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
                 file.write(offset + headerSize, &existGuard,
                            sizeof(existGuard));
             }
+            else if (existGuard.recordId == GUARD_RESOLVED)
```
Please squash this into https://github.com/open-power/guard/pull/2/commits/f8d04388d81dce0c4e1241091e284f68720b2097 (libguard: Changes in deleting the guard record ) in this PR. Looks you are fixing the record delete operation changes but, you added those changes in this PR only so it is not an existing bug. 

You can add those changes details in that commit message.

RameshIyyar commented on guard.cpp:
```diff
@@ -7,19 +7,36 @@
 using namespace std;
 using namespace openpower::guard;
 
+void printHeader(bool display)
```
Please squash this into https://github.com/open-power/guard/pull/2/commits/24a7e7cb5855289200f8f007ac149fc6fbe52c01 (tool: Adding options in guard too) - Looks you are fixing the bug of that patch as a separate commit.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -333,6 +333,52 @@ void clearAll()
+            found = true;
+        }
+        lastPos++;
+    }
```
What happens if no resolved records?

RameshIyyar commented on guard.cpp:
```diff
@@ -7,43 +7,89 @@
 using namespace openpower::guard;
 
-void guardList()
+void printHeader(bool display)
```
Unused function

RameshIyyar commented on guard.cpp:
```diff
@@ -7,43 +7,89 @@
     {
-        if ((elem.errType != GARD_Reconfig) &&
-            (elem.recordId != GUARD_RESOLVED))
+        // Not to print guard records with errorlog type set as GARD_Reconfig
```
Please mention "Why" rather than saying "What doing" - IMO your code itself saying "What doing" (Not to print guard records with errorlog type set as GARD_Reconfig ==  if (elem.errType == GARD_Reconfig)"

RameshIyyar commented on guard.cpp:
```diff
@@ -7,43 +7,89 @@
+        {
+            continue;
+        }
+        // To list resolved records
```
same as above

RameshIyyar commented on guard.cpp:
```diff
@@ -7,43 +7,89 @@
+        {
+            continue;
+        }
+        // To list unresolved records
```
same as above

RameshIyyar commented on guard.cpp:
```diff
@@ -7,43 +7,89 @@
     {
-        std::cout << "No Records to display" << std::endl;
+        std::cout << "No "
+                  << (displayResolved == true ? "resolved" : "unresolved")
```
Do we really need to mention `resolved` vs `unresolved` here? We don't have any records right i.e guard file is empty?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -139,19 +143,33 @@ static GuardRecord getHostEndiannessRecord(const GuardRecord& record)
     {
-        guard_log(GUARD_ERROR, "Size left in GUARD file is %d\n", aval_size);
-        throw GuardFileOverFlowed("Enough size is not available in GUARD file");
+        // Delete the oldest resolved guard record, so that new guard
```
I would prefer this can be handle in create API rather than doing here - IMO we are losing the function scope of "fileSizeCheck"

JS - This can be a separate commit to revert if not required - Your wish

chiragsibm replied:
added changes in create api

RameshIyyar commented on libguard/guard_interface.hpp:
```diff
@@ -63,6 +63,14 @@ void clear(const uint32_t recordId);
+ * @brief Delete the oldest resolved guard record i.e. guard record with
+ *        recordId set as 0xFFFFFFFF.
+ *
+ * @return NULL
```
Add documentation about exceptions that you are throwing.

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -178,10 +178,26 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
-        throw std::runtime_error("Enough size is not available in GUARD file");
+        // Delete the oldest resolved guard record, so that new guard
+        // record can be created.
+        clearResolvedRecord();
```
Looks you are throwing an exception if no resolve record so do you need to differentiate the exception?

chiragsibm replied:
I am adding try catch block here so if there are no resolved records present in the file the clearResolvedRecord will throw an exception saying no resolved records present which i will catch and throw exception as no space left in the file.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -178,10 +178,26 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
-        throw std::runtime_error("Enough size is not available in GUARD file");
+        // Delete the oldest resolved guard record, so that new guard
+        // record can be created.
+        clearResolvedRecord();
```
Please update the commit message about this (why doing this) and here as well.  - I think, it is important to consider so asking to update the commit message and here as well so that we can refer in future.

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -317,6 +333,61 @@ void clearAll()
+        lastPos++;
+    }
+
+    if (found)
```
JS: I think you can do like the below to avoid the `else` block and for more readability.
```
if (!found)
{
        guard_log(GUARD_ERROR, "No resolved Guard record found");
        throw std::runtime_error("No resolved Guard record found");
}

uint32_t offset = delpos * sizeOfGuardRecord;

// To wipe out record from GUARD file
file.erase(offset + headerSize, sizeOfGuardRecord);

// To write the next records to the wiped out location and consecutive
// locations
int i = delpos + 1;
while (lastPos != i)
{
    uint32_t offset1 = i * sizeOfGuardRecord;
    file.read(offset1 + headerSize, &existGuard, sizeOfGuardRecord);
    uint32_t offset2 = (i - 1) * sizeOfGuardRecord;
    file.write(offset2 + headerSize, &existGuard, sizeOfGuardRecord);
    file.write(offset1 + headerSize, &nullGuard, sizeOfGuardRecord);
    i++;
}

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -178,10 +178,26 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
+            lastPos++;
+        }
+
+        avalSize =
```
- I liked your previous patch (You added an internal function to check the file space).
- Do we really need to check file space again here because we cleared the space for one record right, so cannot we create directly?

chiragsibm replied:
Next check is not required. Hence removed it.

RameshIyyar commented on libguard/guard_file.cpp:
```diff
@@ -12,33 +13,26 @@ namespace openpower
-        throw std::runtime_error(
-            "Exception thrown during initialize GUARD file.");
+        guard_log(GUARD_ERROR, "Failed to move to last position in guard file");
+        throw InvalidEntityPath("Exception thrown as failed to move to the "
```
Why throwing `InvalidEntityPath` exception when failed to seek the guard file?

RameshIyyar commented on libguard/guard_file.cpp:
```diff
@@ -12,33 +13,26 @@ namespace openpower
 
 GuardFile::GuardFile(const fs::path& file) : guardFile(file)
 {
-    try
```
Why you are removing the try and catch block for ifstream operation? I think it will throw an exception. Please make sure whether the using ifstream (or its base) class or function will throw an exception or not.

chiragsibm replied:
Changed to original try catch block.

RameshIyyar commented on libguard/guard_file.cpp:
```diff
@@ -57,14 +52,14 @@ void GuardFile::read(const uint64_t pos, void* dst, const uint64_t len)
                   " position= 0x%016llx",
                   pos);
-        throw std::runtime_error("Failed to move to the position in the file");
+        throw InvalidEntityPath("Failed to move to the position in the file");
```
Why throwing `InvalidEntityPath` exception when failed to seek the guard file?

chiragsibm replied:
Added GuardSeekFailed

RameshIyyar commented on libguard/guard_file.cpp:
```diff
@@ -88,15 +83,15 @@ void GuardFile::write(const uint64_t pos, const void* src, const uint64_t len)
                   " Position= 0x%016llx",
                   pos);
-        throw std::runtime_error("Failed to move to the position in the guard file.");
+        throw InvalidEntityPath(
```
Why throwing `InvalidEntityPath` exception when failed to seek the guard file?

chiragsibm replied:
Added GuardSeekFailed.

RameshIyyar commented on libguard/guard_file.cpp:
```diff
@@ -109,7 +104,7 @@ void GuardFile::erase(const uint64_t pos, const uint64_t len)
     {
         guard_log(GUARD_ERROR, "Length passed is %d which is not valid", len);
-        throw std::runtime_error("Not a valid length value");
+        throw InvalidEntityPath("Not a valid length value");
```
Why throwing `InvalidEntityPath` exception when failed to erase in the guard file?

chiragsibm replied:
Added InvalidEntry.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -75,7 +77,7 @@ const fs::path& getGuardFilePath()
     if (guardFilePath.empty())
     {
-        throw std::runtime_error(
+        throw InvalidEntityPath(
```
Why throwing `InvalidEntityPath` exception here? Looks libguard initialization does not happen right?

chiragsibm replied:
Added InvalidEntry

RameshIyyar commented on guard.cpp:
```diff
@@ -7,43 +7,76 @@
+    if (!isHeaderPrinted)
+    {
+        std::cout << "No "
+                  << (displayResolved == true ? "resolved" : "unresolved")
```
JS: `displayResolved == true ?` -> `displayResolved ?`

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -75,7 +77,7 @@ const fs::path& getGuardFilePath()
     if (guardFilePath.empty())
     {
-        throw std::runtime_error(
+        throw InvalidEntry(
```
I think you can throw `GuardFileOpenFailed` which is more appropriate instead of `InvalidEntry` for this interface API.

It will be great if you added a trace as well like an exception message which is throwing.

chiragsibm replied:
Changed to GuardFileOpenFailed and added trace "Guard file not initialized"

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -75,7 +77,8 @@ const fs::path& getGuardFilePath()
     if (guardFilePath.empty())
     {
-        throw std::runtime_error(
+        guard_log(GUARD_ERROR, "Guard file is not initialised.");
```
Do you think adding `"Please make sure libguard_init() is called already"` will help it more from debugging point of view?

chiragsibm replied:
Not really as the user can get the exception. Also with trace itself we can makeout that file was not initialized.

ojayanth commented on guard.cpp:
```diff
@@ -18,7 +18,8 @@ void guardList()
     for (const auto& elem : records)
     {
-        if (elem.errType != GARD_Reconfig)
+        if ((elem.errType != GARD_Reconfig) &&
```
Comments : commit message 
- Based on this commit message,Command line not getting option to clear all guard records. . This will be use for lab/debug purpose.

chiragsibm replied:
As per discussion changed the clearall option and marking all records as resolved. User can erase one by one resolved records. You want me to change this to earlier version i.e. to do a real clear of the guard records or add a new option to be used for lab purpose?

chiragsibm replied:
As per discussion adding new option to erase the guard records.

ojayanth replied:
Clearall user expectation is to clear all guard record from memory . here , you are ot clearing instead , changing to resolved. Istead of changing existing behavior , better to use new option to change the state. This will avoid changes in the existing tool user and automation scripts.

chiragsibm replied:
We are trying to make guard tool same as in FSP, as previously we were not invalidating the resources, we are correcting that behavior in this PR.
New guard options:

Usage: guard [OPTIONS]

Options:
  -h,--help                   Guard CLI tool options
  -c,--create TEXT            Create Guard record, expects physical path as input
  -i,--invalidate TEXT        Invalidate a single Guard record, expects physical path as input
  -I,--invalidate-all         Invalidates all the Guard records
  -l,--list                   List all the Guard'ed resources
  -r,--reset                  Erase all the Guard records
  -v,--version                Version of GUARD tool

ojayanth commented on README.md:
```diff
@@ -34,14 +34,16 @@ GUARD Tool
+  -d,--delete TEXT         Delete GUARD record, expects physical path as input.
+  -l,--list                Listing of GUARDed resources.
+  -r,--clearall            Clears GUARD states for all resources.
+  -s,-clearresolvedrecord  Delete GUARD states for oldest resolved resource
```
is this option intend to clear only the oldest one ?

chiragsibm replied:
Yes, as per discussion made this change and clearing the oldest guard record, this is basically to not lose the history of resolved records.

ojayanth replied:
That is production use case , am refering guard tool usecase.

chiragsibm replied:
This option is removed because of the changes in clearResolvedRecord function i.e.
Hostboot removes the resolved/invalid guard record and writes the new record in the same location. Hence providing this as an option in tool can create problem. Hence removing it.

RameshIyyar replied:
This option did not implement in this patch. - You forget remove?

ojayanth commented on README.md:
```diff
@@ -34,14 +34,16 @@ GUARD Tool
+  -c,--create TEXT         Create GUARD record, expects physical path as input.
+  -d,--delete TEXT         Delete GUARD record, expects physical path as input.
+  -l,--list                Listing of GUARDed resources.
+  -r,--clearall            Clears GUARD states for all resources.
```
This is just marks records resolved?  not a real clear.  How the user can clear all records permanently?

chiragsibm replied:
Yes, this option will now mark every guard record as resolved. Nope user cannot do an erase(real clear) of records in one go. He can erase/real clear only one record i.e. resolved record at a given time.

ojayanth replied:
The rquiremnet you are refering is production use case , agreed that . 
Am refering the Gurad tool use case , which is mainly used by lab/developers/internal teams . This is mandatory requirement from tool point to support clear all the guard records.  Looks like you are deviating the base requirement here.

chiragsibm replied:
Providing new option to erase the guard records, this option i.e -r will mark every record present as invalid/resolved.

ojayanth replied:
keep the existing "-r, --clearall", clearAll, , which should clear all the records from flash , use diffrent options to support new options. This is to support  backward-compatibility and support exitsing automations infrstructre.

chiragsibm replied:
Made the changes.

ojayanth commented on libguard/guard_interface.cpp:
```diff
@@ -317,6 +327,59 @@ void clearAll()
     }
 }
 
+void clearResolvedRecord()
```
The logic seems to be not good fro  flash performance. , need to look better option to identify the oldest record instead of position based approch. Is this same logic is used by hostboot code?

chiragsibm replied:
Not sure of hostboot code. I saw they mark every record as resolved when do clear, not sure how they are handling this. We don't have any time stamp to check that's why went with this approach. May be later can see if something better can be done here.

ojayanth replied:
Still prefer to avoid unnecessary flash right , for each delete operation. incase if this is not requirement for existing hostboot point , need to relook this flow.

chiragsibm replied:
Hostboot removes the resolved/invalid guard record when no space is left in file to create a new record and writes the new record in that empty slot. So this method will not do unnecessary flash write.  Opting for this method in libguard as well.
Note: In this case the record ID will come in random order.

dhruvibm commented:
Looks good to me

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -167,9 +171,20 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
+        // Storing the 1st resolved guard record position
+        if (existGuard.recordId == GUARD_RESOLVED)
+        {
+            if (!isResolvedRecordPresent)
```
Cannot we merge the inner `if` block with the outer `if` block?

chiragsibm replied:
Yes we can merge it.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -167,9 +171,20 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
+            if (!isResolvedRecordPresent)
+            {
+                empPos = pos;
+                isResolvedRecordPresent = true;
```
Do we really need this variable? 
Just we cannot use `empPos`? empPos will be **non-zero** if resolved record is present else **zero**
- If can, can you change the name of the variable for `empPos`?

chiragsibm replied:
Suppose there are multiple invalid records so every time the empPos will get changed, we want 1st invalid record location/position. Any better idea please feel free to share.

RameshIyyar replied:
I think you can initialize the `empPos` with `-1` (we can not use `0` because there may be a chance that the first record can be resolved - I guess If you have a header then `0` we won't get it) and check with `-1` before assigning the first resolved position so that we can avoid that `isResolvedRecordPresent` variable i.e we can use `empPos == -1` instead of `!isResolvedRecordPresent`.

chiragsibm replied:
Done.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -178,17 +193,32 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
 
     GuardRecord guard;
     //! if blank record exist
     if (isBlankRecord(existGuard))
```
When this case will occur? Just looking for the code which you added in this block for this patch.

chiragsibm replied:
This case will not happen, as the size check we are doing above.

RameshIyyar replied:
JS: Do you want to remove it then? OR it will impact something if removed this check.

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -178,17 +193,32 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
-        offset = lastPos * sizeof(guard);
+        // If oldest resolved record is deleted, than new record will be
+        // written in that empty slot
+        if (resolveRecordClear)
```
Cannot we use the LN 196 (new code) block to set `offset`?

chiragsibm replied:
I prefer adding it here rather in LN 196 to avoid confusion.

RameshIyyar replied:
OK, I would prefer in LN196 so that we can avoid one more `if and else` block and also, IMO when we are doing something different based on the condition then that can go into the same block statement (if possible) to give more readability.  

Any advantage to prefer this way?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -167,9 +171,20 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
+            }
+        }
+
+        id = be32toh(existGuard.recordId);
```
Why you are preferring a new variable for this instead of using like previous code? Looks `id` is used in the `if` block alone.

chiragsibm replied:
Yes, its required, as i was seeing issues with recordId of new records.

RameshIyyar replied:
It will be great if you mention the issue details in the comment section while saying some issues so that reviewer can understand. - Nevermind :)

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -167,9 +171,20 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
             return getHostEndiannessRecord(existGuard);
         }
+
+        // Storing the 1st resolved guard record position
```
- I think this statement is not correct because we may have the record (which is matched with the given record id) that can be resolved and you are skipping that right?
- Do we need to use the record slot (which is matched with the given record id and that is resolved) to create a new record if no space in the GUARD file? What is the behavior in HostBoot for the same case?

chiragsibm replied:
Why its wrong, I am storing the position not the record ID. I am not getting please elaborate. 
libguard is doing same as hostboot.

RameshIyyar replied:
Let's assume (we can store max 3 records) like below.

```
FFFFFFFF core0
00000001 core1
FFFFFFFF core2
```
Now trying to guard a `core0` then which slot you will consider?  1st slot or 3rd slot? If you think, 3rd then the statement  `Storing the 1st resolved guard record position` is correct?

(I think you won't get 1st slot position since we are trying the same entity (core0) which is already resolved so it moves to the next record based on the above code logic) - correct me if I am wrong. 

If My guess is correct then, I like to know the behavior of the same case in the HostBoot side when you mentioned HB will use the first resolved record slot - IMO, we need to use the first slot (assume the above example) instead of 3rd slot if space is full since user/application trying to create a record for the entity which is already resolved. - Considering guard history use case.

I guess I am thinking very corner case :)

chiragsibm replied:
I meant 1st invalid/resolved record encountered so in the above case FFFFFFFF core0 will be used.
It doesnt matter if the entity path is same its the oldest invalid record present.

RameshIyyar replied:
> I meant 1st invalid/resolved record encountered so in the above case FFFFFFFF core0 will be used

How if trying to create `core0`?
https://github.com/chiragsibm/guard/commit/033591a341216f2af5f5994851a8f1156f036261#diff-afe02f82ae6baea2b83e14a0385bd9715fc2cd75ba556e86e68bd87984d40d09R163  - It won't go into the next record?

> It doesnt matter if the entity path is same its the oldest invalid record present.

I guess, it matters from the guard history use case if you are using the core2 instead of using the core1 slot.

chiragsibm replied:
I meant oldest invalid/resolved record.
As per above example "FFFFFFFF core0" should be replaced.
Did the changes :)

RameshIyyar commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,83 @@
+  public:
+    explicit GuardException(const std::string& message) : message(message){};
+
+    virtual const char* what() const noexcept override
```
Why you are using both **virtual** and **override** keywords ? The `what()` is virtual function in `std::exception` class.

RameshIyyar commented on guard.cpp:
```diff
@@ -90,6 +123,8 @@ int main(int argc, char** argv)
         std::optional<std::string> deleteGuardStr;
         bool listGuardRecords = false;
         bool clearAll = false;
+        bool clearResolvedRecord = false;
```
Unused variable
- did you ran clang?
- did you enable support in the meson.build to treat warnings as errors?

chiragsibm replied:
I removed this in next commit in this PR.

RameshIyyar replied:
Please remove in the same commit - Now you can ignore. The reviewer will add comments unnecessarily by looking at the current commit/patch :)

RameshIyyar commented on README.md:
```diff
@@ -51,6 +53,9 @@ Physical path formats supported by guard tool:-
 * /sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
 * sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
 
+**Note:** clearresolvedrecord option will delete the oldest/first resolved
```
This option did not implement in this patch. - You forget remove?

chiragsibm replied:
Yes forgot to remove this from README.

RameshIyyar commented on guard.cpp:
```diff
@@ -141,6 +146,8 @@ int main(int argc, char** argv)
         app.add_flag("-p, --listresolvedrecords", listResolvedGuardRecords,
                      "Listing of resolved GUARDed resources");
+        app.add_flag("-e, --eraseall", eraseall,
+                     "Erase all the GUARDed resources");
```
- Add the proper difference between `-e` vs `-r` in the help section and README. 
  - I think in English `clear` and `erase` will have the same meaning so the user may get confused.
- Do you need to communicate about the new changes of the clear option with all stakeholders (whoever using the guard tool in their component? For example MFG, CI, TestTeam, etc) on this?
- (Just checking) Do we need an option to just delete (remove from GUARD file) alone all resolved records?

chiragsibm replied:
Will add the details in README. Yes, we need to communicate to test team. Once this is merged will do that.
I don't think there is any requirement to just remove invalid guard records.

RameshIyyar commented on README.md:
```diff
@@ -34,14 +34,14 @@ GUARD Tool
-  -v,--version           Version of GUARD tool.
+  -h,--help                Guard CLI tool options
+  -c,--create TEXT         Create Guard record, expects physical path as input
+  -i,--delete TEXT         Invalidate a single Guard record, expects physical
```
Looks old, please change it.

RameshIyyar commented on README.md:
```diff
@@ -51,6 +52,10 @@ Physical path formats supported by guard tool:-
 * /sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
 * sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
 
+**Note:**
```
Looks you did not change here i.e using as old
Also, you used bullet but, not showing as a bullet - please correct if planning to change.

ojayanth commented on libguard/guard_interface.cpp:
```diff
@@ -136,6 +136,7 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
     int lastPos = 0;
     uint32_t maxId = 0;
     uint32_t offset = 0;
+    uint32_t avalSize = 0;
```
Commit message : please provide details , why this changes are done now.

chiragsibm replied:
libguard: Adding space check before creating record

Changes:
-Adding check in create API, which will calculate the available
size in the file before writing the guard record.

ojayanth commented on libguard/guard_interface.cpp:
```diff
@@ -169,6 +170,14 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
+    avalSize = file.size() - ((lastPos + 1) * sizeof(existGuard) + headerSize);
+	if (avalSize < sizeof(existGuard))
+    {
+        guard_log(GUARD_ERROR, "Size left in GUARD file is %d\n", avalSize);
```
Noticed some issues with file size calculation, debug point include file.size information.

chiragsibm replied:
Added to capture file.size() in trace.

ojayanth commented on guard.cpp:
```diff
@@ -18,7 +18,8 @@ void guardList()
     for (const auto& elem : records)
     {
-        if (elem.errType != GARD_Reconfig)
+        if ((elem.errType != GARD_Reconfig) &&
```
Commit message: Change the commit message , "Add support for guard record history".

General direction:  Good to see the options , which is not creating much impact on existing tool usage.
Am little to confused with the way you organised the code to implement this support now.
"-Change in clearAll function to mark recordIds of guard records
as FF..FF" -- Make sure this is not going to break the existing guard -r optionin this commit.  update commit message accordingly.

chiragsibm replied:
Made changes. Done.

ojayanth commented on libguard/guard_interface.cpp:
```diff
@@ -231,127 +237,82 @@ GuardRecords getAll()
+    uint32_t offset = 0;
 
-    memset(&guard, 0, sizeof(guard));
+    memset(&existGuard, 0, sizeof(existGuard));
```
This is going to break the existing -r option right?  need to make sure re-order the commits with out impacting the existing functionality.
High level you should be are adding feature on top of existibg one , like , here adding new option to invalidate  instead of delete. and clearall you are planing to keep both options right?

chiragsibm replied:
Made the changes accordingly. -r usecase will not break.
Created invalidateAll function not changing clearAll.

ojayanth commented on libguard/guard_interface.cpp:
```diff
@@ -137,21 +137,31 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType)
     uint32_t offset = 0;
```
commit message : "libguard: Changes in create api " please provide meaningfull subject. here looks ore generic one.

chiragsibm replied:
Done.

ojayanth commented on README.md:
```diff
@@ -34,14 +34,14 @@ GUARD Tool
-  -l,--list              Listing of GUARDed resources.
-  -r,--clearall          Clears GUARD states for all resources.
-  -v,--version           Version of GUARD tool.
+  -h,--help                Guard CLI tool options
```
Commit message : Add  details "why this changes required?" 
based on your initial commit message comments  -r is not changing to -I , it is juts changing -clear-all option, looks like you are removing -r option here , contradicyting the initail commmit response. ,update accordingly , refder precious  comments .  Stopping the review now, will continue receview , once you clarified the direction .

chiragsibm replied:
Done.

ojayanth commented on guard.cpp:
```diff
@@ -18,7 +18,8 @@ void guardList()
     {
-        if (elem.errType != GARD_Reconfig)
+        if ((elem.errType != GARD_Reconfig) &&
+            (elem.recordId != GUARD_RESOLVED))
```
Not seeing the update in commit message related to this change.

chiragsibm replied:
Done.

ojayanth commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,83 @@
+{
+namespace exception
+{
+// TODO:Issue #3325, to avoid multiple exception classes
```
should be creating issue in guard repository.  
TODO #3325 is useless unless you specify the repository name .

chiragsibm replied:
Created issue in guard repo. 
https://github.com/open-power/guard/issues/3

ojayanth commented on libguard/guard_interface.cpp:
```diff
@@ -234,127 +240,74 @@ GuardRecords getAll()
-}
-
-void clear(const EntityPath& entityPath)
+static void invalidateRecord(const EntityPath& entityPath,
```
This seems to error prone. you are assuming one of the value is good here. incase user provided both arguments then error path is an issue. Also not providing usefull debug data here for failure case.
Since no0 header in the function, am assuming you are expecting only one parameter at a time . to achive this you need to decalre as stad::variant and handle properly.

Otherwise use the existing flow in the code used for deleteRecord().

chiragsibm replied:
using std::variant. Made the changes.

ojayanth replied:
if the record pos starts with 0 , then the logic is broken here . if that is the case intilize with FFFFFFFF. so that the logic holds good.

chiragsibm replied:
No, record position i.e. recordId can never be 0. We start it with 1. If we initialize with FFFF..FF than it will try to invalidate the already invalidated guard record(because FFFFFFFF is a recordId in guard i.e. invalidated guard record)

ojayanth commented on libguard/guard_common.hpp:
```diff
@@ -1,17 +1,21 @@
 #include <cstdint>
-#include <stdexcept>
 #include <optional>
+#include <stdexcept>
```
General comment: all the exported api's ( in guard_interface.hpp) should updated with possible exception list. instead of making generic statement.  "  Throw exception on failure".

chiragsibm replied:
As a practice applying this to all the functions throwing custom exceptions.

ojayanth commented on README.md:
```diff
@@ -34,14 +34,15 @@ GUARD Tool
+  -l,--list                List all the GUARD'ed resources
+  -r,--reset               Erase all the Guard records
+  -v,--version             Version of GUARD tool
+
```
- a  details option is missing here.

chiragsibm replied:
Not added -a details as it is hidden in guard tool.
In readme only adding the options shown in help text(i.e. guard -h)

ojayanth commented on libguard/guard_exception.hpp:
```diff
@@ -0,0 +1,83 @@
+namespace exception
+{
+// TODO:Issue #3, to avoid multiple exception classes
+class GuardException : public std::exception
```
Exception list provided in the exported interfaces need to revisit , not seeing bunch of exception related file operations here in the futnction hedare definition

chiragsibm replied:
We are handling file open, seek, read and write error. Can you please share which are missed ?
Mostly file exceptions are thrown in file operations i.e. guard_file.cpp

ojayanth replied:
These guard fle  helper API's are consumed  exported interfaces right, You should look that also.

chiragsibm replied:
Added the changes. Done

RameshIyyar reviewed (approved):
LGTM - Tested with HardwareIsolation manager to make sure actual libguard is working or not.
